src/qcodes/parameters/infer.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, ClassVar
+
+from qcodes.instrument import Instrument, InstrumentBase, InstrumentModule
```

Comment:
The qcodes.instrument module depends on the qcodes on the qcodes.parameters module. 

Thinking about this more generally it feels like these convenience functions really should be methods on the parameter and or instrument classes but I realize that it may take some time to iron that out. Therefor I think I would suggest to add this to the qcodes.extensions module for now.